SalesforceCommerceCloud/pwa-kit
# [Localization] Set the correct currency when switching locales

GUS: https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE000005MF7MYAW/view

# Description

This PR implements the specific association of locales and currencies represented in the table for all the Commerce API requests that accept the `currency` parameter.

| Locale  | Currency |
| ------------- | ------------- |
| en-GB  | GBP |
| fr-FR  | EUR |
| it-IT  | EUR |
| zh-CN | CNY |
| ja-JP  | JPY |

The preferred currencies for each langue are defined in the `package.json` file:
https://github.com/SalesforceCommerceCloud/pwa-kit/blob/a1ba9329d7dbec1aab5e947d508e50d97a6217cc/packages/pwa/package.json#L186-L217


The PR also refactors:
- Categories context Provider by adding a new `<CategoriesProvider />` and a new `useCategory()` hook.
- The util function `buildUrlLocale` now called `getUrlWithLocale`.
---
These are some of the API calls that include the `currency` value as a query parameter or in the body of the request:

## PLP
**PENDING SCAPI FIX**: The `currency` query parameter sent to the `product-search` endopint is ignored. The response always uses the default currency `"currency": "GBP"`. [Slack thread](https://salesforce-internal.slack.com/archives/C02B75CCFV2/p1631300673055200).

#### SCAPI [`ShopperSearch`](https://salesforcecommercecloud.github.io/commerce-sdk-isomorphic/classes/shoppersearch.shoppersearch-1.html) 
Methods: [`productsearch`](https://salesforcecommercecloud.github.io/commerce-sdk-isomorphic/classes/shoppersearch.shoppersearch-1.html#productsearch).

## PDP
#### SCAPI [`ShopperProducts`](https://salesforcecommercecloud.github.io/commerce-sdk-isomorphic/classes/shopperproducts.shopperproducts-1.html) 
Methods: [`getProduct`](https://salesforcecommercecloud.github.io/commerce-sdk-isomorphic/classes/shopperproducts.shopperproducts-1.html#getproduct) and [`getProducts`](https://salesforcecommercecloud.github.io/commerce-sdk-isomorphic/classes/shopperproducts.shopperproducts-1.html#getproducts).

## Cart/Checkout/Add to cart
The Basket is created using the method `createBasket` of the `OcapiShopperBaskets` class.

#### OCAPI [`Baskets`](https://documentation.b2c.commercecloud.salesforce.com/DOC1/topic/com.demandware.dochelp/OCAPI/current/shop/Resources/Baskets.html) resource
| HTTP method  | Resource |
| ------------- | ------------- |
| POST  | [/baskets](https://documentation.b2c.commercecloud.salesforce.com/DOC1/topic/com.demandware.dochelp/OCAPI/current/shop/Resources/Baskets.html#iid-1036386844__id49484514) |

The POST request creates a new basket initialized with default values. This method apparently doesn't allow us to customize the `currency` value on the body of the request.

I added a new method `updateBasket` to the `OcapiShopperBaskets` class that updates the basket currency.
The method is called when the current currency doesn't match the current locale.
https://github.com/SalesforceCommerceCloud/pwa-kit/blob/2ba4f636eda0ff9da05f19ae129c82f785abccc9/packages/pwa/app/commerce-api/ocapi-shopper-baskets.js#L28

The `updateBasket` method uses the OCAPI:
#### OCAPI [`Baskets`](https://documentation.b2c.commercecloud.salesforce.com/DOC1/topic/com.demandware.dochelp/OCAPI/current/shop/Resources/Baskets.html) resource
| HTTP method  | Resource |
| ------------- | ------------- |
| PATCH  | [/baskets/{basket_id}](https://documentation.b2c.commercecloud.salesforce.com/DOC1/topic/com.demandware.dochelp/OCAPI/current/shop/Resources/Baskets.html#id-1036386844__id1441479317) |





# Types of Changes




> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes
- Added currencies for each locale.
- Added currency parameter to API calls.
- Added `updateBasket` method to the `OcapiShopperBaskets` class to updates the basket currency.
- Refactor `useCategory` and `useCurrency` hooks and contexts providers `<CategoriesProvider />` and `<CurrencyProvider />`.
- Fixed and added tests

# How to Test-Drive This PR

1. Navigate across all the pages that show a currency value (Home, PLP, PDP, Cart, Checkout, My account, etc.)
2. Verify that the currency matches the selected locale as per the association of locales and currencies in the table.






## General


## Accessibility Compliance

You must check off all items in **one** of the follow two lists:


_or..._


## Localization
